### PR TITLE
Add specs for Integer ceil, floor, truncate and round that takes optional digits

### PR DESCRIPTION
**Pull request by boddhisattva**
This is currently Work in Progress(WIP). 

**Update**: This is now complete based on the requirement specified as part of [this issue](https://github.com/ruby/spec/issues/473).

### CODE REVIEW

**Comment by boddhisattva** [comment]
Not sure if I'm missing something wrt the Integer `ceil` and `floor` function  but it looks like as per the ruby 2.4.0 docs for [ceil](https://ruby-doc.org/core-2.4.0/Integer.html#method-i-ceil) and [floor](https://ruby-doc.org/core-2.4.0/Integer.html#method-i-floor) return an Integer.

1. `1.ceil(2)        #=> 1`
2. `1.floor(2)        #=> 1`

But in Ruby 2.4.0 console it looks like they return a Float:

```ruby
2.4.0 :036 > 1.ceil(2)
 => 1.0
2.4.0 :037 > 1.floor(2)
 => 1.0
```

If the docs need to be updated, I can try to submit a patch to fix this in the [ruby github repo](https://github.com/ruby/ruby) as per the contribution guidelines specified [here](https://bugs.ruby-lang.org/projects/ruby/wiki/HowToContribute) but before I do that, can one please confirm on the above observation and also please confirm if I'm using the correct approach to update the docs? Is my understanding correct that the docs get picked up from [here](https://github.com/ruby/ruby/blob/6b05153a3a75b74b64553d6a46f501d9ee0f0376/numeric.c#L5076-L5093) for `Integer#ceil` and [here](https://github.com/ruby/ruby/blob/6b05153a3a75b74b64553d6a46f501d9ee0f0376/numeric.c#L5044-L5061) for `Integer#floor` ?

**Comment by eregon** [comment]
Your observations are right:
```ruby
$ ruby -v
ruby 2.4.1p111 (2017-03-22 revision 58053) [x86_64-linux]
[5] pry(main)> 42.round 2
=> 42.0
[6] pry(main)> 42.ceil 2
=> 42.0
[7] pry(main)> 42.floor 2
=> 42.0
[8] pry(main)> 42.truncate 2
=> 42.0
```

Actually, this behavior changed in 2.5 (see https://github.com/ruby/spec/pull/436), but remains as you observe in 2.4.
It seems the problem comes from ruby-doc.org, which seems to take documentation from trunk and not 2.4.2 as it should. I commented on https://github.com/documenting-ruby/ruby/issues/81#issuecomment-335005854
Here is the source for 2.4.2: https://github.com/ruby/ruby/blob/v2_4_2/numeric.c#L5060

So ideally we should specify the behavior for both versions ranges like in https://github.com/ruby/spec/pull/436/files#diff-839feea4a131eb7231be5b63a9d0654cR7.

File: core/integer/ceil_spec.rb [hunk 1]
```diff
@@ -3,4 +3,25 @@
...
+      it "returns the smallest integer greater than self with at least precision.abs trailing zeros" do
+        18.ceil(-1).should eql(20)
+        18.ceil(-2).should eql(100)
+        18.ceil(-3).should eql(1000)
+        -1832.ceil(-1).should eql(-1830)
+        -1832.ceil(-2).should eql(-1800)
+        -1832.ceil(-3).should eql(-1000)
+      end
+    end
+  end
```
Comment (eregon) [inline core/integer/ceil_spec.rb]: This looks good! :smiley:

Comment (boddhisattva) [reply core/integer/ceil_spec.rb]: Thanks @eregon :smiley:

**Comment by boddhisattva** [comment]
@eregon Interesting, what does the `trunk` branch signify? Sorry, I'm not quite aware of this naming convention of branches as part of the `Ruby` organization of projects. Also, what version of Ruby should the `trunk` branch ideally point to at any given point of time ?

I realise specs for `Integer#round` have already been added [here](https://github.com/ruby/spec/blob/master/core/integer/round_spec.rb#L78-L94). I've finished with the set of specs for `ceil`, `floor` and `truncate` as part of those that need to be wrapped up wrt Ruby 2.4 based on the issue specified [here](https://github.com/ruby/spec/issues/473).  Can the specs for 2.5 wrt `ceil`, `floor` and `truncate` be treated as a separate PR that anyone can take up ?

**Comment by eregon** [comment]
Subversion's `trunk` is like git's `master`, i.e. the development branch (so currently what will become 2.5 but also development for the next versions).

**Comment by eregon** [comment]
Yes, of course the 2.5 part can be done separately.
Thank you for the new specs!

**Comment by eregon** [comment]
(FWIW I fixed the 2.5 behavior in 64c06f9c9faccba0c461fd15eccaaa2df04dacfc, not so trivial indeed)

**Comment by boddhisattva** [comment]
Thanks for collaborating on this and sharing more info @eregon :) . I didn't know that they were using `trunk` in the context of subversion.

Thanks for adding the specs for Ruby 2.5 👍  . I was thinking of attempting it later if feasible but glad that you already have done them.

I'm curious about one thing in general, it would be great to get more insights to better understand the value of these specs. I mean I do know that Ruby 2.4 is already made available to end users but some specs here are still for some 2.4 features that are already made available to everyone then what additional value in general is there with adding these specs now ? 

Also, I believe that whenever releasing a new ruby version for everyone's use, they would already have some sort of specs for the new features and bug fixes right and if so, where do those tests ideally reside ? I guess that is a different repo or place from what we have here ?

Thanks for your time answering my questions.

**Comment by eregon** [comment]
> What additional value in general is there with adding these specs now ?
> Also, I believe that whenever releasing a new ruby version for everyone's use, they would already have some sort of specs for the new features and bug fixes right and if so, where do those tests ideally reside ? I guess that is a different repo or place from what we have here ?

These are very good questions, with unfortunately not such a simple answer.

Tests are added when new features are added, such as https://github.com/ruby/ruby/commit/a1542d33ca30cdecb741be526d043be6bba01d76 in the case of Integer#ceil.
These specific tests look good and in this case might even have better coverage than the specs we wrote. But there are major differences which makes the specs here worthwhile:
* These testes are written by the implementer of the feature, so it's basically a list of cases the author thought about. Writing tests from another perspective often discovers new edge cases (it's not so rare to find bugs while writing specs).
* As you see MRI unit tests are just assertions with close to zero comments. For more complex methods, it becomes very hard to understand why the method should behave in such a way or should satisfy those assertions. Having descriptions of the behavior makes the specs a lot easier to read, but also to implement without needing to always look at MRI's code (which for example in this case is very tied to C arithmetic semantics and might be irrelevant if this is re-implemented in another language).
* These tests live in the MRI repository, which means when the behavior changes, the tests are just changed in place. This means it can be very confusing (as you saw) when behavior changes and there is no good place to look at. ruby/spec really shines here, it's very clear [what is the behavior for which Ruby version](https://github.com/ruby/spec/blob/487a79d19ce2e0b6a9be6106d6be0b56dd87fda6/core/integer/shared/integer_rounding.rb).
* MRI tests tend to be specific to MRI. The organization of MRI tests is sometimes rather surprising (try finding a do/while loop test). MRI tests sometimes rely on advanced features like refinements in the test harness. These are a few issues that make it hard (but not impossible) for other implementations to reuse those tests, and much more so for early implementations.
* There are other reasons, for instance mentioned [here](http://blade.nagaokaut.ac.jp/cgi-bin/scat.rb/ruby/ruby-core/78633) but I do want to make this too long.

Now, I am trying to encourage MRI developers adding new features to write specs, and this would be very valuable as they are expert about the feature since they implement it. We made progress so at least [much](https://bugs.ruby-lang.org/issues/13156) [less](https://bugs.ruby-lang.org/issues/13666) technical issues are standing in the way. But writing specs is different than writing unit tests, and it's hard to change habits so this might take a while. The current policy in MRI is you can write MRI tests or ruby/spec specs for a new feature/bug fix, and developers are free to choose what they prefer.
However, this is something that only happened during the 2.5 timeline so for 2.4 specs we basically need to catch up.

**Comment by boddhisattva** [comment]
Wow! It's amazing to see the depth to which you have answered my question @eregon . Thank you very much :) I know better understand the value of writing these specs. Some food for thought may be, Do you think this specific repository could benefit with a **Motivation** section in the README in terms of why writing these specs is encouraged ? Hopefully, that may give people more insights on why they should consider contributing to this repo and may be also help them better realise the value they're bringing to the community through their contributions here ? 

This was like the bigger picture for me and I definitely found it really insightful. Thanks once again for the detailed explanation Benoit :)
